File: Task Tracker Java/contest_java/contestNetworking/src/main/java/com/example/protobuff/ProtoUtils.java
package com.example.protobuff;

import com.example.domain.Participant;
import com.example.domain.ParticipantTask;
import com.example.domain.User;
import com.example.dto.TaskDTO;
import com.example.dto.ParticipantDTO;
import com.example.enums.Type;
import com.example.enums.AgeGroup;

import java.util.ArrayList;
import java.util.List;

public class ProtoUtils {
    public static ContestProto.ContestRequest createLoginRequest(User user) {
        ContestProto.User loggedInUser = ContestProto.User.newBuilder().setId(user.getId()).setUsername(user.getUsername()).setPassword(user.getPassword()).build();
        return ContestProto.ContestRequest.newBuilder().setReq(ContestProto.ContestRequest.Req.Login).setUser(loggedInUser).build();
    }

    public static ContestProto.ContestRequest createLogoutRequest(User user) {
        ContestProto.User loggedOutUser = ContestProto.User.newBuilder().setId(user.getId()).setUsername(user.getUsername()).setPassword(user.getPassword()).build();
        return ContestProto.ContestRequest.newBuilder().setReq(ContestProto.ContestRequest.Req.Logout).setUser(loggedOutUser).build();
    }

    public static ContestProto.ContestRequest createGetTasksEnrolledRequest() {
        return ContestProto.ContestRequest.newBuilder().setReq(ContestProto.ContestRequest.Req.GET_ALL_TASKS_EXPERIMENTAL).build();
    }

    public static ContestProto.ContestRequest createGetFilteredParticipantsRequest(Type type, AgeGroup ageGroup) {
        String[] filters = new String[2];
        filters[0] = String.valueOf(type);
        filters[1] = String.valueOf(ageGroup);
        return ContestProto.ContestRequest.newBuilder().setReq(ContestProto.ContestRequest.Req.FILTER).setType(filters[0]).setAge(filters[1]).build();
    }

    public static ContestProto.ContestRequest createRegisterRequest(String name, int age, Type task1, Type task2) {
        String type1 = String.valueOf(task1);
        String type2 = String.valueOf(task2);
        ContestProto.ParticipantTask participantTask = ContestProto.ParticipantTask.newBuilder().setName(name).setAge(age).setType1(type1).setType2(type2).build();
        return ContestProto.ContestRequest.newBuilder().setReq(ContestProto.ContestRequest.Req.ADD_PARTICIPANT_EXPERIMENTAL).setPt(participantTask).build();
    }

    public static ContestProto.ContestResponse createOKResponse() {
        return ContestProto.ContestResponse.newBuilder().setRep(ContestProto.ContestResponse.Reply.OK).build();
    }

    public static ContestProto.ContestResponse createErrorResponse(String text) {
        return ContestProto.ContestResponse.newBuilder().setRep(ContestProto.ContestResponse.Reply.Error).setError(text).build();
    }

    public static ContestProto.ContestResponse createGetTasksEnrolledResponse(List<ContestProto.TaskDTO> tasks) {
        ContestProto.ContestResponse.Builder response = ContestProto.ContestResponse.newBuilder().setRep(ContestProto.ContestResponse.Reply.GET_ALL_TASKS_EXPERIMENTAL);
        for (var task : tasks) {
            ContestProto.TaskDTO taskDTO = ContestProto.TaskDTO.newBuilder().setTaskID(task.getTaskID()).setType(task.getType()).setAge(task.getAge()).setEnrolled(task.getEnrolled()).build();
            response.addTasks(taskDTO);
        }
        return response.build();
    }

    public static ContestProto.ContestResponse createGetFilteredParticipantsResponse(List<ContestProto.ParticipantDTO> participantDTOS) {
        ContestProto.ContestResponse.Builder response = ContestProto.ContestResponse.newBuilder().setRep(ContestProto.ContestResponse.Reply.FILTER);
        for (var pd : participantDTOS) {
            ContestProto.ParticipantDTO pdto = ContestProto.ParticipantDTO.newBuilder().setName(pd.getName()).setAge(pd.getAge()).build();
            response.addParticipants(pdto);
        }
        return response.build();
    }

    public static ContestProto.ContestResponse createRegisterResponse() {
        return ContestProto.ContestResponse.newBuilder().setRep(ContestProto.ContestResponse.Reply.ADD_PARTICIPANT_EXPERIMENTAL).build();
    }

    public static ContestProto.ContestResponse createUpdateResponse(){
        return ContestProto.ContestResponse.newBuilder().setRep(ContestProto.ContestResponse.Reply.UPDATE).build();
    }

    public static String getError(ContestProto.ContestResponse response) {
        return response.getError();
    }

    public static User getUser(ContestProto.ContestRequest request){
        return new User(request.getUser().getUsername(), request.getUser().getPassword());
    }

    public static ParticipantTask getPT(ContestProto.ContestRequest request){
        return new ParticipantTask(request.getPt().getName(), request.getPt().getAge(), Type.valueOf(request.getPt().getType1()), Type.valueOf(request.getPt().getType2()));
    }

    public static String[] getFilters(ContestProto.ContestRequest request){
        String[] filters = new String[2];
        filters[0] = request.getType();
        filters[1] = request.getAge();
        return filters;
    }

    public static List<ContestProto.TaskDTO> getEnrolledTasks(ContestProto.ContestResponse response) {
        List<ContestProto.TaskDTO> tasks = new ArrayList<>();
        for(int i = 0; i < response.getTasksCount(); i++){
            tasks.add(response.getTasks(i));
        }
        return tasks;
    }

    public static List<ContestProto.TaskDTO> getEnrolledTasksForResponse(List<TaskDTO> tasks){
        List<ContestProto.TaskDTO> res = new ArrayList<>();
        for(var tk: tasks){
            ContestProto.TaskDTO td = ContestProto.TaskDTO.newBuilder().setTaskID(tk.getTaskID()).setType(ContestProto.Type.valueOf(tk.getType().toString())).setAge(ContestProto.AgeGroup.valueOf(tk.getAgeGroup().toString())).setEnrolled(tk.getEnrolled()).build();
            res.add(td);
        }
        return res;
    }

    public static List<ContestProto.Participant> getFilteredPartiticipants(List<Participant> parts){
        List<ContestProto.Participant> res = new ArrayList<>();
        for(var p: parts){
            ContestProto.Participant pt = ContestProto.Participant.newBuilder().setId(p.getId()).setName(p.getName()).setAge(p.getAge()).build();
            res.add(pt);
        }
        return res;
    }

    public static List<ParticipantDTO> part2dto(List<Participant> pts){
        List<ParticipantDTO> res = new ArrayList<>();
        for(var p: pts){
            res.add(new ParticipantDTO(p.getName(), p.getAge()));
        }
        return res;
    }

    public static List<ContestProto.ParticipantDTO> getFilteredPartsDTOMapper(List<ContestProto.Participant> pts) {
        List<ContestProto.ParticipantDTO> parts = new ArrayList<>();
        for(var p: pts){
            parts.add(ContestProto.ParticipantDTO.newBuilder().setName(p.getName()).setAge(p.getAge()).build());
        }
        return parts;
    }

    public static List<ContestProto.ParticipantDTO> getFilteredPartsDTO(ContestProto.ContestResponse response) {
        List<ContestProto.ParticipantDTO> parts = new ArrayList<>();
        for(int i = 0; i < response.getParticipantsCount(); i++){
            parts.add(response.getParticipants(i));
        }
        return parts;
    }
}

File: Task Tracker C#/contest/contestNetworking/ProtoUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Com.Example.Networking;
using MyType = contestDomain.enums.Type;
using proto = Com.Example.Networking;
using System.Reflection.Metadata.Ecma335;
using System.Globalization;

namespace contestNetworking
{
    static class ProtoUtils
    {
        public static ContestRequest CreateLoginRequest(contestDomain.User user)
        {
            proto.User user1 = new proto.User { Id = user.ID, Username = user.Username, Password = user.Password };
            ContestRequest request = new ContestRequest { Req = ContestRequest.Types.Req.Login, User = user1 };
            return request;
        }

        public static ContestRequest CreateLogoutRequest(contestDomain.User user)
        {
            return new ContestRequest { Req = ContestRequest.Types.Req.Logout, User = new User { Id = user.ID, Username = user.Username, Password = user.Password } };
        }

        public static ContestRequest CreateTasksEnrolledRequest()
        {
            return new ContestRequest { Req = ContestRequest.Types.Req.GetAllTasksExperimental };
        }

        public static ContestRequest CreateGetFilteredParticipantsRequest(MyType type, contestDomain.enums.AgeGroup age)
        {
            string[] filters = new string[2] { nameof(type), nameof(age) };
            return new ContestRequest { Req = ContestRequest.Types.Req.Filter, Type = filters[0], Age = filters[1] };
        }

        public static ContestRequest CreateRegisterRequest(string name, int age, MyType task1, MyType task2)
        {
            proto.ParticipantTask pt = new proto.ParticipantTask { Name = name, Age = age, Type1 = task1.ToString(), Type2 = task2.ToString() };
            return new ContestRequest { Req = ContestRequest.Types.Req.AddParticipantExperimental, Pt = pt };
        }

        public static ContestResponse CreateOKResp()
        {
            return new ContestResponse { Rep = ContestResponse.Types.Reply.Ok };
        }

        public static ContestResponse CreateErrorResponse(string text)
        {
            return new ContestResponse { Rep = ContestResponse.Types.Reply.Error, Error = text };
        }

        public static ContestResponse CreateGetTasksEnrolledResponse(List<proto.TaskDTO> tasks)
        {
            ContestResponse response = new ContestResponse { Rep = ContestResponse.Types.Reply.GetAllTasksExperimental };
            foreach (var task in tasks)
            {
                proto.TaskDTO taskDTO = new proto.TaskDTO { TaskID = task.TaskID, Type = task.Type, Age = task.Age, Enrolled = task.Enrolled };
                response.Tasks.Add(taskDTO);
            }
            return response;
        }

        public static ContestResponse CreateGetFilteredParticipantsResponse(List<proto.ParticipantDTO> parts)
        {
            ContestResponse response = new ContestResponse { Rep = ContestResponse.Types.Reply.Filter, Participants = { parts } };
            /*foreach (var participant in parts)
            {
                response.Participants.Add(participant);
            }*/
            return response;
        }

        public static ContestResponse CreateRegisterResponse()
        {
            return new ContestResponse { Rep = ContestResponse.Types.Reply.AddParticipantExperimental };
        }

        public static ContestResponse CreateUpdateResponse()
        {
            return new ContestResponse { Rep = ContestResponse.Types.Reply.Update };
        }

        public static string GetError(ContestResponse response)
        {
            return response.Error;
        }

        public static contestDomain.User GetUser(ContestRequest request)
        {
            return new contestDomain.User(request.User.Id, request.User.Username, request.User.Password);
        }

        public static contestDomain.ParticipantTasks GetPT(ContestRequest request)
        {
            Console.WriteLine(request.Pt.Type1.ToString());
            return new contestDomain.ParticipantTasks(request.Pt.ParticipantID, request.Pt.Name, request.Pt.Age, (MyType)Enum.Parse(typeof(MyType), request.Pt.Type1.ToString()), (MyType)Enum.Parse(typeof(MyType), request.Pt.Type2.ToString()));
        }

        public static List<TaskDTO> GetTasksConverted(List<contestDomain.dto.TaskDTO> tasks)
        {
            List<TaskDTO> res = new List<TaskDTO>();
            foreach (var tk in tasks)
            {
                TextInfo textInfo = new CultureInfo("en-US", false).TextInfo;
                string title = textInfo.ToTitleCase(tk.type.ToString().ToLower());
                string ageG = textInfo.ToTitleCase(tk.ageGroup.ToString().ToLower());
                TaskDTO taskDTO = new TaskDTO { TaskID = tk.taskID, Type = Enum.Parse<proto.Type>(title), Age = (proto.AgeGroup)Enum.Parse(typeof(proto.AgeGroup), ageG), Enrolled = tk.Enrolled };
                res.Add(taskDTO);
            }
            return res;
        }

        public static string[] GetFilters(ContestRequest request)
        {
            return new string[2] { request.Type, request.Age };
        }

        public static List<Participant> GetFilteredParticipantsConverted(List<contestDomain.Participant> participants)
        {
            List<Participant> res = new List<Participant>();
            foreach (var participant in participants)
            {
                Participant p = new Participant { Id = participant.ID, Name = participant.Name, Age = participant.Age };
                res.Add(p);
            }
            return res;
        }

        public static List<ParticipantDTO> GetFilteredParticipantsDTO(List<contestDomain.Participant> participants)
        {
            List<ParticipantDTO> res = new List<ParticipantDTO>();
            foreach (var participant in participants)
            {
                ParticipantDTO pd = new ParticipantDTO { Name = participant.Name, Age = participant.Age };
                res.Add(pd);
            }
            return res;
        }

        public static List<TaskDTO> GetEnrolledTasks(ContestResponse response)
        {
            List<TaskDTO> tasks = new List<TaskDTO>();
            for (int i = 0; i < response.Tasks.Count; i++)
            {
                tasks.Add(response.Tasks.ElementAt(i));
            }
            return tasks;
        }

        public static List<ParticipantDTO> GetFilteredPartsDTO(ContestResponse response)
        {
            List<ParticipantDTO> parts = new List<ParticipantDTO>();
            for (int i = 0; i < response.Participants.Count; i++)
            {
                parts.Add(response.Participants.ElementAt(i));
            }
            return parts;
        }
    }
}
